eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/location/googleapiclient/GoogleApiClientHolder.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright 2017 SchedJoules
+ *
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.schedjoules.eventdiscovery.framework.location.googleapiclient;
+
+import android.support.v4.app.FragmentActivity;
+
+import com.google.android.gms.common.api.GoogleApiClient;
+import com.schedjoules.eventdiscovery.framework.utils.Listenable;
+
+
+/**
+ * A component holding a {@link GoogleApiClient.Builder} initially to allow multiple services to be set up, in other words multiple client components to
+ * configure it independently. Then once it is built, the clients can access the same single instance of {@link GoogleApiClient}.
+ * <p>
+ * Note: setting {@link GoogleApiClient.OnConnectionFailedListener} on this holder is different from setting it on the builder or the apiclient, the former
+ * listens to errors that were not resolvable by the auto-managed client, while the latter listens to all errors (triggered even if auto-manage is handling it).
+ * See {@link GoogleApiClient.Builder#enableAutoManage(FragmentActivity, int, GoogleApiClient.OnConnectionFailedListener)} for more.
+ *
+ * @author Gabor Keszthelyi
+ */
+public interface GoogleApiClientHolder extends Listenable<GoogleApiClient.OnConnectionFailedListener>
```

Comment:
Long story short: I looked into this and started to refactor but I suggest that we create a separate technical task for this, so that this PR can be merged.

I've pushed those changes to separate branch: `stories/134-error-handling-google-api-facade'
The complexity there is with the callbacks. At this point I am still not sure if everything will work correctly with this dynamically recreated GoogleApiClients, so I would like to get back to that and review, but please have a first look at it yourself as well. The interface looks like this
```
public interface GoogleApiClientFacade
{
    void addApi(@NonNull Api<? extends Api.ApiOptions.NotRequiredOptions> googleApi);

    void addConnectionFailedListener(GoogleApiClient.OnConnectionFailedListener listener);

    void registerConnectionCallback(GoogleApiClient.ConnectionCallbacks connectionCallbacks, boolean forceCreate);

    PendingResult<AutocompletePredictionBuffer> getAutocompletePredictions(String query, AutocompleteFilter filter);

    PendingResult<PlaceBuffer> getPlaceById(String placeId);

    @RequiresPermission(anyOf = { "android.permission.ACCESS_COARSE_LOCATION", "android.permission.ACCESS_FINE_LOCATION" })
    Optional<Location> getLastLocation();
}
```

I am quite confident that the current version in this PR doesn't have risk in how this `GoogleApiClientHolder` works, the temporal coupling there is intended and used correctly now. I agree though that it's not good design like this as you pointed out, for example the client could be built with the builder directly now, which would be a different instance. But as it is used now, it should be fine, so I think merging is fine.